styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java

```diff
@@ -135,85 +139,95 @@ void tick0() {
   private void triggerAndProgress(Backfill backfill) {
     // Do not include all backfill spans in parent tick span to avoid it growing too big
     tracer.spanBuilderWithExplicitParent("Styx.BackfillTriggerManager.triggerAndProgress", null)
-        .startSpanAndRun(() -> triggerAndProgress0(backfill));
+        .startSpanAndRun(() -> Try.run(() -> triggerAndProgress0(backfill)));
   }
 
-  private void triggerAndProgress0(Backfill backfill) {
-    final Optional<Workflow> workflowOpt;
-    try {
-      workflowOpt = storage.workflow(backfill.workflowId());
-    } catch (IOException e) {
-      LOG.warn("Failed to read workflow {}", backfill.workflowId(), e);
-      return;
-    }
+  private void triggerAndProgress0(Backfill backfill) throws IOException {
 
-    if (!workflowOpt.isPresent()) {
-      LOG.debug("workflow not found for backfill {}, halt it.", backfill);
-      storeBackfill(backfill.builder().halted(true).build());
+    // Check how many instances for this backfill are currently active
+    // Note: getActiveStatesByTriggerId is eventually consistent and may not return all triggered instances.
+    var alreadyActiveInstances = stateManager.getActiveStatesByTriggerId(backfill.id()).size();
+
+    // Look up the workflow. Halts the backfill if the workflow does not exist.
+    var workflowOpt = readBackfillWorkflowOrHalt(backfill.id());
+    if (workflowOpt .isEmpty()) {
       return;
     }
+    var workflow = workflowOpt.orElseThrow();
 
-    final Workflow workflow = workflowOpt.get();
+    for (int i = 0; i < MAX_INSTANCES_TICK_TRIGGER; i++) {
```

Comment:
So you're capping the number of instances to run per backfill per tick. Is that to make sure that bigger backfills don't starve others? Any other reason?

Reply:
Non-starvation is one reason. Another is that the `alreadyActiveInstances` might get stale.

Reply:
Added a comment.